apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/helper/DubboTraceHelper.java

```diff
@@ -0,0 +1,162 @@
+        if (version != null && version.length() > 0) {
+            requestName += " version=" + version;
+        }
+
+        return requestName;
+    }
+
+    @VisibleForAdvice
+    public static Span createConsumerSpan(DubboApiInfo apiInfo, InetSocketAddress remoteAddress) {
+        TraceContextHolder<?> traceContext = DubboTraceHelper.tracer.getActive();
+        if (traceContext == null) {
+            return null;
+        }
+        Span span = traceContext.createExitSpan();
+        if (span == null) {
+            return null;
+        }
+
+        span.withType(EXTERNAL_TYPE)
+            .withSubtype(DUBBO_SUBTYPE)
+            .withName(buildDubboRequestName(apiInfo));
+        Destination destination = span.getContext().getDestination();
+        destination.withAddress(remoteAddress.getHostName()).withPort(remoteAddress.getPort());
+
+        Destination.Service service = destination.getService();
+        service.withType(EXTERNAL_TYPE).withResource(DUBBO_SUBTYPE).withName(DUBBO_SUBTYPE);
+
+        return span.activate();
+    }
+
+    @VisibleForAdvice
+    public static void fillTransaction(Transaction transaction, DubboApiInfo dubboApiInfo) {
+        transaction.withName(buildDubboRequestName(dubboApiInfo));
+        transaction.withType("dubbo");
+        transaction.activate();
+    }
+
+    public static boolean isBizException(Class<?> interfaceClass, Class<?> exp) {
+        String apiJarFile = getJarFile(interfaceClass);
+        if (apiJarFile == null) {
+            return false;
+        }
+        return apiJarFile.equals(getJarFile(exp));
+    }
+
+    public static String getJarFile(Class<?> clazz) {
+        if (clazz == null) {
+            return null;
+        }
+        ProtectionDomain domain = clazz.getProtectionDomain();
+        if (domain == null) {
+            return null;
+        }
+        CodeSource source = domain.getCodeSource();
+        if (source == null) {
+            return null;
+        }
+        URL location = source.getLocation();
+        if (location == null) {
+            return null;
+        }
+        return location.getFile();
+    }
+
+    public static void doCapture(Object[] args, Throwable t, Object returnValue) {
+        Transaction transaction = tracer.currentTransaction();
+        if (transaction == null) {
+            return;
+        }
+        boolean hasError = t != null;
+        CoreConfiguration coreConfig = tracer.getConfig(CoreConfiguration.class);
+        CoreConfiguration.EventType captureBody = coreConfig.getCaptureBody();
+        if (CoreConfiguration.EventType.OFF.equals(captureBody) ||
+            (CoreConfiguration.EventType.ERRORS.equals(captureBody) && !hasError)) {
+            return;
+        }
+
+        captureArgs(transaction, args);
+        if (t != null) {
+            transaction.addCustomContext("throw", t.toString());
+        } else {
+            transaction.addCustomContext("return", returnValue.toString());
```

Comment:
Custom context is reserved for user-defined properties. Core agent plugins should not use them.

Reply:
sometimes i want to capture the arguments and return object, including exception thrown, and show them in kibana, where should i store them in?

Reply:
Would it be possible to register an interceptor that is custom to your application that adds the custom context?

Reply:
Like `capture_body` for http request, it is quite common to capture something for dubbo request，for example, the params, the data was return or the exception thrown.

The difference is the dubbo agent store these data into the `custom context`. 

Is it proper to store these data into `SpanContext` or `TransactionContext`, like http?

Reply:
There's currently no proper field for that. I agree that could be useful to store parameters and return values but it's not Dubbo specific. I also applies to gRPC, for example.
So I suggest leaving this out in this PR and open a follow-up discussion in https://github.com/elastic/apm instead, according to the guidelines in https://github.com/elastic/apm/blob/master/docs/agents/new-fields.md. This makes sure we align across agents and that the proposal is not Dubbo or Java specific.

> or the exception thrown.

We capture the transaction via `transaction.captureException` now, right?

Reply:
ok， i will remove this out of this PR

Reply:
It has been removed now.